llvm/docs/CASObjectFormats/FlatV1.md

```diff
@@ -0,0 +1,83 @@
+# FlatV1: a "flat"/simple CAS-optimized object format
+
+"FlatV1" is a CAS-optimized object format with a flat hierarchy.
+
+The top-level CAS object is called a *compile-unit*. Descriptions of
+symbols/blocks/atoms are outlined into leaf CAS objects and the graph is stored
+in data at the top level.
+
+This format is called "flat" because there is no nesting; the only CAS object
+that directly references other objects is compile-unit.
+
+The code is in:
+- llvm/include/llvm/CASObjectFormats/FlatV1.h
+- llvm/lib/CASObjectFormats/FlatV1.cpp
+
+See also: CommonEncodings.md.
+
+## Major caveats with current state
+
+This format has a monolithic top-level node that describes most things. This
+node never repeats when an object file changes, so whats inside it cannot be
+deduped. This format is smaller than opaque Mach-O and it has much better
+growth rate, but its growth rate is not as good as NestedV1.
+
+For example, we've looked at data for storing 100 builds of consecutive commits
+of LLVM. The storage cost overhead for 100 builds vs. 1 build is:
+- release builds
+    - Opaque: 56%
+    - FlatV1: 23%
+    - NestedV1: 13%
+- no-opt builds (`-O0 -gnone`)
+    - Opaque: 74%
+    - FlatV1: 28%
+    - NestedV1: 11%
+
+See also "Major caveats" section in CASObjectFormats.md.
+
+## Structure
+
+The top-level CAS object is called *compile-unit*. This top-level object is
+effectively the whole object file, with description templates for various
+concepts outlined to reduce aggregate storage cost.
+
+**compile-unit**: all graph information stored at the top-level
+- Data: all symbols, blocks, sections, referring to templates with a 4B index
```

Comment:
Compile-unit in flatv1 is basically:
* Template array: CASIDs to other CAS objects that stores Symbol, Blocks, etc.
* Data: an array of indexes that encode all the references from one entities to another. The array works based on the fact that there is a deterministic traversal order to visit all the sections, symbols and blocks and each of them know how many indexes they need when visiting.

Reply:
Thanks; cleaned up the description a bit!